Add unsubscribe hook and saga

huwshimi opened the pull request:
## Done

- Add support for unsubscribing to the websocket sagas.
- Add clean up and unsub to fetch and get hooks.

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the machine list.
- Look in `machine` in the redux store and you should see the `lists` requests.
- Navigate away from the machine list, preferably to somewhere that doesn't have machines.
- You should see the dispatches for `machine/unsubscribe` and state should no longer have the lists requests.
- Go to a machine details page and navigate away.
- You should see the dispatches for `machine/unsubscribe` and state should no longer have the details requests.

## Fixes

Fixes: canonical/app-tribe#1141.
Fixes: canonical/app-tribe#1128.

webteam-app commented:
Demo starting at https://maas-ui-4319.demos.haus

petermakowski commented:
On machine listing, sometimes only half of `machine.lists` are removed.

To reproduce: 
1. Go to machine listing page
2. refresh the page
3. navigate away to the settings page

On the machine details page all of `machine.details` are removed correctly.

huwshimi commented:
> On machine listing, sometimes only half of `machine.lists` are removed.

What was happening is that when the machine list loaded the fetch hook was running twice when the filters changed from `undefined` to `{}` which was causing a clean up of the first request and a dispatch for the second request, but I was being defensive in the reducers and creating the entry in the store if the response came back but the call couldn't be found in the store. However, because we were making two calls in quick succession the call came back after it was cleaned up which resulted in orphaned entries in the store.

I've fixed `{}` and falsy filters being equal so that the request only happens once and also fixed the orphaned entries issue. That all might more sense when you see the code :).
